### ISSUE TITLE
mileage, engine, max_power and torque are never parsed to numeric, so they get one-hot encoded

### ISSUE THREAD

**Issue by sabhishek13-py** [shyamsivamanicam-cmd/carprice-ml#1]
## Summary

In the CarDekho dataset, `mileage`, `engine`, `max_power` and `torque` arrive as strings with units attached (`23.4 kmpl`, `1248 CC`, `74 bhp`, `190Nm@ 2000rpm`). Nothing in `train.py` or `notebook.ipynb` strips those units, so `get_preprocessor` routes all four columns to the one-hot branch instead of treating them as continuous numbers.

The result is that the four most predictive vehicle specs are encoded as ~1,300 indicator columns, and only 4 features are genuinely numeric.

## Evidence

The notebook's own `df.info()` output (cell 3) confirms the dtypes:

```
 8   mileage        7907 non-null   str
 9   engine         7907 non-null   str
 10  max_power      7913 non-null   str
 11  torque         7906 non-null   str
```

`get_preprocessor` splits columns purely by dtype (train.py:58-59):

```python
cat_cols = X.select_dtypes(include='object').columns.tolist()
num_cols = X.select_dtypes(include=np.number).columns.tolist()
```

Running the pipeline on a CarDekho-shaped frame gives:

```
treated as CATEGORICAL: ['fuel', 'seller_type', 'transmission', 'owner',
                         'mileage', 'engine', 'max_power', 'torque']
treated as NUMERIC    : ['km_driven', 'seats', 'car_age', 'km_per_year']
```

Cardinalities from cell 5 of the notebook: mileage 393, engine 121, max_power 322, torque 441.

## Why this matters

The models are fitting one-hot indicators for individual power and torque strings rather than learning from continuous engine specs. Two consequences visible in the committed notebook output:

- Ridge wins with R² 0.9591, ahead of XGBoost at 0.9526. Linear regression beating gradient boosting on tabular data is backwards, and is what you would expect when a very wide sparse matrix lets the model memorise near-unique spec strings.
- Any spec value not seen in training hits `handle_unknown='ignore'` and encodes as all zeros, so the model falls back to the intercept for unseen cars. That is the common case at inference time.

## Suggested fix

Parse the units in `feature_engineering` before the split, for example:

```python
df['mileage']   = df['mileage'].str.extract(r'([\d.]+)').astype(float)
df['engine']    = df['engine'].str.extract(r'([\d.]+)').astype(float)
df['max_power'] = df['max_power'].str.extract(r'([\d.]+)').astype(float)
```

Two details worth deciding on:

- `mileage` mixes `kmpl` and `km/kg` units across fuel types, so the raw number is not comparable across rows. Either normalise it or keep the unit as a separate categorical feature.
- `torque` needs its own handling. It encodes two quantities in one string, in inconsistent units (`190Nm@ 2000rpm` vs `12.7@ 2,700(kgm@ rpm)`). Splitting it into a torque value plus an RPM value, with kgm converted to Nm, is the usual approach. Dropping it is also reasonable, but right now it is neither parsed nor dropped, and it is not listed in the `required_cols` check in train.py:39 or in the README feature list.

Expect the model ranking to change once this is fixed, which means the `best_name == 'XGBoost'` gate around hyperparameter tuning in train.py:175 should be revisited at the same time.
